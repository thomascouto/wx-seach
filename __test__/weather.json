{
	"coord": {
		"lon": -87.65,
		"lat": 41.85
	},
	"weather": [
		{
			"id": 804,
			"main": "Clouds",
			"description": "overcast clouds",
			"icon": "04n"
		}
	],
	"base": "stations",
	"main": {
		"temp": 2.76,
		"feels_like": 2.76,
		"temp_min": 1.29,
		"temp_max": 3.63,
		"pressure": 1020,
		"humidity": 96
	},
	"visibility": 145,
	"wind": {
		"speed": 0.45,
		"deg": 51,
		"gust": 1.79
	},
	"clouds": {
		"all": 100
	},
	"dt": 1706187259,
	"sys": {
		"type": 2,
		"id": 2018264,
		"country": "US",
		"sunrise": 1706188185,
		"sunset": 1706223336
	},
	"timezone": -21600,
	"id": 4887398,
	"name": "Chicago",
	"cod": 200,
	"dt_txt": ""
}