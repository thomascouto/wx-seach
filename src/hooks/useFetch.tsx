import { useState } from 'react';

import { getWeatherService } from '@/services';

export default function useFetch<T>() {
  const [weather, setWeather] = useState<T>();
  const [error, setError] = useState<ErrorResponse>();
  const [isLoading, setIsLoading] = useState(false);

  const getWeatherData = ({ q, isForecast, units = 'metric' }: QueryProps) => {
    setIsLoading(true);
    getWeatherService<T>({
      q,
      units,
      isForecast,
    })
      .then(setWeather)
      .catch((err: string) => {
        const parsedError = JSON.parse(err) as ErrorResponse;
        setError(parsedError);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return {
    getWeatherData,
    weather,
    isLoading,
    error,
  };
}
