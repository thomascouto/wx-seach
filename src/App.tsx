import { lazy } from 'react';

import { HistoryCard, Spinner, TextInput } from '@/components';
import useFetch from '@/hooks/useFetch';

import styles from './App.module.scss';

const Card = lazy(() => import('@/components/Card'));

function App() {
  const { getWeatherData, weather, isLoading } = useFetch();

  const handleData = (value: string, units: Units) => {
    getWeatherData(value);
  };

  return (
    <>
      <nav className={styles.navigation}>
        <HistoryCard theme="sunny" />
      </nav>
      <section className={styles.searchBar}>
        <TextInput action={handleData} />
      </section>
      <section className={styles.weatherSection}>
        {isLoading ? <Spinner /> : weather && <Card data={weather} />}
      </section>
    </>
  );
}

export default App;
