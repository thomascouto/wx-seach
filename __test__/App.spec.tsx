import weather from './weather.json';

import { render, screen } from '@testing-library/react';

import App from '@/App';
import * as useFetch from '@/hooks/useFetch';

describe('App tests', () => {
  it('should render correctly', () => {
    jest.spyOn(useFetch, 'default').mockImplementation(() => ({
      getWeatherData: jest.fn(),
      weather,
      isLoading: false,
      error: null,
    }));
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });

  it('should render an error message', () => {
    jest.spyOn(useFetch, 'default').mockImplementation(() => ({
      getWeatherData: jest.fn(),
      weather: null,
      isLoading: false,
      error: { cod: 404, message: 'city not found' },
    }));

    const { container } = render(<App />);

    const h3 = screen.getByRole('heading');

    expect(h3.innerHTML).toBe('Error! city not found');
    expect(container).toMatchSnapshot();
  });
});
