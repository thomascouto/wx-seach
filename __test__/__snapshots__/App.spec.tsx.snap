// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should render an error message 1`] = `
<div>
  <section
    class="searchBar"
  >
    <div
      class="inputContainer"
    >
      <input
        class="textInput"
        placeholder="Search for a location..."
        type="text"
        value=""
      />
      <div
        class="toggleButtons"
      >
        <button
          class="active"
        >
          째C
        </button>
        <button
          class=""
        >
          째F
        </button>
      </div>
    </div>
  </section>
  <section
    class="error"
  >
    <h3>
      Error! 
      city not found
    </h3>
  </section>
</div>
`;

exports[`App tests should render correctly 1`] = `
<div>
  <section
    class="searchBar"
  >
    <div
      class="inputContainer"
    >
      <input
        class="textInput"
        placeholder="Search for a location..."
        type="text"
        value=""
      />
      <div
        class="toggleButtons"
      >
        <button
          class="active"
        >
          째C
        </button>
        <button
          class=""
        >
          째F
        </button>
      </div>
    </div>
  </section>
  <section
    class="weatherSection"
  />
</div>
`;
